ARCs/arc-0023.md

```diff
@@ -0,0 +1,76 @@
+---
+arc: 23
+title: Sharing Application's information
+description: Append informations to the compiled application's bytes
+author: Stéphane Barroso (@sudoweezy)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/80
+status: Draft
+type: Meta
+created: 2023-01-11
+---
+
+## Abstract
+
+The following document introduces a convention for appending information to the compiled application's bytes. 
+The goal of this convention is to standardize the process of verifying and interacting with smart contracts. The encoded byte is `arc23` followed by the length of the CID representing a folder of files & the CID itself.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative.
+
+### CID
+The <a href="https://github.com/multiformats/cid">CID</a> allows to access the corresponding files using <a href="https://docs.ipfs.tech/">IPFS</a>.
+
+The CID **MUST** represent one file or a folder of files.
+The minimum requirement for this ARC is a file representing the ABI (ABI as described in [ARC-4](arc-0004.md))
+
+### Encoding
+The encoded byte is `arc23` followed by the length of the CID, the number of files, and the CID itself.
+
+eg: for the example stored in `/asset/arc-0023/Folder`, 
+`arc23 59 2 bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici`
+`bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici` correspond to this Folder: `../assets/arc-0023/Folder`. It can be obtained with `ipfs add --cid-version=1 Folder/*  -q -w -n| tail -1`
+
+If there is only one file, the CID **MUST** be related to the ABI of the application. 
+eg: `arc23 59 1 bafkreiajguyi3i5hf4h7blslsv53rjbexzzvpkyhdf26hmuqthrk4iko4a`
+`bafkreiajguyi3i5hf4h7blslsv53rjbexzzvpkyhdf26hmuqthrk4iko4a` correspond to this Abi File: `../assets/arc-0023/Folder/Abi.json`. It can be obtained with `ipfs add --cid-version=1 Folder/Abi.json -q`
+
+> The option -n is for obtaining the hash without pinning the file, you **MAY** remove the option in production
+
+Otherwise, if there is more than one file, the CID must correspond to a folder of files. The ABI **MUST** be accessible by querying and IPFS node with `{cid}/Abi.json`
+eg:`bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici/Abi.json`.
+
+### Appending
+The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing 1-byte constant which is the encoded object.
```

Comment:
```suggestion
The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing `[]byte` constant which is the encoded object.
```

Reply:
There are 3 properties we want of the way the information is included in the program:

1. Easy to extract: it should be easy to detect when ARC-23 information are added and extract them. In particular, this should not require the ability to fully parse TEAL code, nor find boundaries of op-codes. The number of false positive should be negligible and the code to parse should be easy enough.
2. Easy to securely add to programs: it should be easy to add to new programs in a fully automated and secure way. This means that adding it should not break security. In addition, the TEAL compiler should be able to handle this gracefully. See also https://github.com/algorandfoundation/ARCs/pull/80#issuecomment-1252662213
3. Short: it should not be too long.

If we enforce that the CID is always a V1 CID w/ SHA256, this makes it easier because then

```
bytecblock b64(YXJjMjMBcBIg48nKgteefr5CesSEPH/gc55s1HzVG40dsN6NpCHjKBI=)
```

is always encoded as (in hexadecimal):

`0x2601296172633233` followed by 36 bytes of the CID

(`0x26` is the `bytecblock` opcode, `0x01` is the number of []byte strings, `0x29` is the length = 41 bytes, `0x6172633233` is hexadecimal of `arc23)

If in addition it's always supposed to be at the end of the TEAL code, then it's very easy to parse:
go to 44 bytes before end of file.

Now, an alternative would be:

```
bytecblock "arc23" b64(CID)
```

is easier to parse in case the CID may have different length since then you can look for `0x2602056172633233` as prefix that is then followed by a varuint of the length of the CID and the CID.
This is easier than concatenating "arc23" to the CID and having a single []byte bytecblock element because we have a very easy long prefix (otherwise, it can get messy).

But if the length of the CID is fixed, then the first option saves 1 byte.